### PR TITLE
ARROW-4493: [Rust] Make accumulate_scalar somewhat exhaustive and easier to read

### PR DESCRIPTION
**Pull request by tupshin**
redoing patch that was originally against the non-apache repo

### CODE REVIEW

**Comment by nevi-me** [comment]
Hi @tupshin, can you please run `cargo fmt` on your PR to fix formatting errors.

File: rust/datafusion/src/execution/aggregate.rs [hunk 1]
```diff
@@ -246,42 +246,42 @@ impl AggregateFunction for SumFunction {
...
+                (Int16(a), Int16(b)) => Some(Int16(*a + b)),
+                (Int32(a), Int32(b)) => Some(Int32(*a + b)),
+                (Int64(a), Int64(b)) => Some(Int64(*a + b)),
+                (Float32(a), Float32(b)) => Some(Float32(*a + b)),
+                (Float64(a), Float64(b)) => Some(Float64(*a + b)),
+
+
+                (Utf8(_), _) => panic!("Cannot SUM() Utf8 values"),
+                (Struct(_), _) => panic!("Cannot SUM() structs"),
+                (Float32(_), _) => panic!("Cannot SUM() Float32 values"),
```
Comment (nevi-me) [inline rust/datafusion/src/execution/aggregate.rs]: Please update the wording to "Cannot SUM() Float32 with non Float32 values", same with `Float64`

File: rust/datafusion/src/execution/aggregate.rs [hunk 2]
```diff
@@ -246,42 +246,42 @@ impl AggregateFunction for SumFunction {
...
-                }
-                (Some(ScalarValue::Float64(a)), Some(ScalarValue::Float64(b))) => {
-                    Some(ScalarValue::Float64(a + *b))
-                }
-                _ => panic!("unsupported data type for SUM"),
-            }
+        use ScalarValue::*;
+        self.value = match (&self.value, value) {
+            (None, _) => value.clone(), 
+            (Some(__), None) => unimplemented!(),
```
Comment (nevi-me) [inline rust/datafusion/src/execution/aggregate.rs]: The `unimplemented!()` here, especially without context, doesn't look good. I can see that the current code leaves that code path, so I'm curious as to what would happend if `value` is `None`. Was it panicking?

**Review by nevi-me — CHANGES_REQUESTED** [review]
@andygrove 's review will trump mine, but so far I think you could do more on this PR by either:

* making the same change to other instances of `fn accumulate_scalar()`
* wrapping the function in a macro, so you could use the macro for `MAX` and `MIN`

**Comment by andygrove** [comment]
Hi @tupshin I'm going to be reviewing this PR in the next day or two. Sorry for the delay.

**Comment by andygrove** [comment]
@tupshin Thanks for cleaning this up a little. If we are modifying this code though we should change this method to return `Result` rather than have calls to `panic!` or `unimplemented!`.

**Comment by nevi-me** [comment]
Hi @andygrove @tupshin I believe this PR might be redundant. Andy cleaned up a lot of the datafusion code just before 0.13. I'd vote that we close this, we have a few Rust PRs that are getting stale.

**Comment by andygrove** [comment]
@nevi-me I agree. Closing this one.

**Pull request closed without merging**